File: src/main/cpp/subsystems/intake/IntakeRotate.cpp
#include "subsystems/intake/IntakeRotate.h"


IntakeRotate::IntakeRotate() {
    // Setup current limit, idle mode, and encoder factor
    rotation.SetSmartCurrentLimit(
            CONSTANTS::INTAKE::ROTATION::HARD_CURRENT_LIMIT.to<int>());
    rotation.SetSecondaryCurrentLimit(
            CONSTANTS::INTAKE::ROTATION::SOFT_CURRENT_LIMIT.to<double>());
    encoder.SetPositionConversionFactor(
            CONSTANTS::INTAKE::ROTATION::TURN_TO_DEGREE.to<double>());
            // The encoder is now in units of degrees

    // Set up PID controller
    pid.SetP(CONSTANTS::INTAKE::ROTATION::PID::P);
    pid.SetI(CONSTANTS::INTAKE::ROTATION::PID::I);
    pid.SetD(CONSTANTS::INTAKE::ROTATION::PID::D);
    pid.SetFF(CONSTANTS::INTAKE::ROTATION::PID::FF);
    pid.SetIZone(CONSTANTS::INTAKE::ROTATION::PID::I_ZONE);
    pid.SetOutputRange(CONSTANTS::INTAKE::ROTATION::PID::MIN,
            CONSTANTS::INTAKE::ROTATION::PID::MAX);
}

units::degree_t IntakeRotate::Get() {
    return angle;
}

void IntakeRotate::Set(units::degree_t angle) {
    // Save the angle
    IntakeRotate::angle = angle;

    // Update the PID controller
    pid.SetReference(angle.to<double>(),
            rev::CANSparkMaxLowLevel::ControlType::kPosition);
}

units::degree_t IntakeRotate::GetActual() {
    return (units::degree_t) encoder.GetPosition();
}

void IntakeRotate::Zero() {
    encoder.SetPosition(0.0);
}

File: src/main/java/frc/bionicpolars/subsystems/intake/IntakeRotate.java
package frc.bionicpolars.subsystems.intake;

import com.revrobotics.CANSparkMax;
import com.revrobotics.RelativeEncoder;
import com.revrobotics.SparkMaxPIDController;

import frc.bionicpolars.Constants;

public class IntakeRotate {
    public static final double GEARBOX_RATIO = 100.0,
            PULLEY_RATIO = 122.0 / 65.0,
            TURN_TO_DEGREE = 360 / (GEARBOX_RATIO * PULLEY_RATIO),
            P = 0.07,
            I = 0.0,
            D = 0.0;
    private CANSparkMax motor = new CANSparkMax(
            Constants.INTAKE_ROTATION_ID,
            CANSparkMax.MotorType.kBrushless);
    RelativeEncoder encoder = motor.getEncoder();
    SparkMaxPIDController pid = motor.getPIDController();
    double angle = 0;


    public IntakeRotate() {
        encoder.setPositionConversionFactor(TURN_TO_DEGREE);
        pid.setP(P);
        pid.setI(I);
        pid.setD(D);
        pid.setOutputRange(-1.0, 1.0);
    }

    public double Get() {
        return angle;
    }

    public void Set(double angle) {
        this.angle = angle;
        pid.setReference(angle, CANSparkMax.ControlType.kPosition);
    }

    public double GetActual() {
        return encoder.getPosition();
    }

    public void Zero() {
        encoder.setPosition(0);
    }
}
